### ISSUE TITLE
Compile-time usage of parts of strutils fails when using -d:useNimRtl

### ISSUE THREAD

**Issue by awr1** [nim-lang/Nim#8405]
```nim
import strutils

const a = "abc" in "abcdef"
echo(a)
```
console output:
```
c:/nimtests $ nim c -d:useNimRtl conststr.nim
Hint: used config file 'c:\Nim\config\nim.cfg' [Conf]
Hint: system [Processing]
Hint: conststr [Processing]
Hint: strutils [Processing]
Hint: parseutils [Processing]
Hint: math [Processing]
Hint: bitops [Processing]
Hint: algorithm [Processing]
Hint: unicode [Processing]
lib\pure\strutils.nim(1501, 10) Error: cannot 'importc' variable at compile time
```

Numerous procs in strutils have the same problem, e.g. `contains` and `toUpperASCII`. We should not be trying to link to `nimrtl` in matters of compile-time evaluation (as far as most everything in strutils is concerned).

**Comment by Araq** [comment]
Yeah, I know about this problem too, but I fail to see a solution. If you link dynamically, why can we assume the current version is fine for compiletime evaluation?

**Comment by awr1** [comment]
Is there a way of fixing something like this with `when nimvm`? I took a crack at it but, while it "fixed" the issue, it ended up not producing exports for the functions in `nimrtl` upon closer inspection. Not technically ideal. Also worth mentioning that in the docs it isn't made obvious that `nimrtl` is the complete standard library beyond the GC manager. Perahaps an option to build a stripped-down `nimrtl` - "`mini_nimrtl`" - containing just the GC might be nice? IDK.

Currently I have a personal project I've been working on that uses a fork of the standard library with the `{.rtl.}` pragma removed in `strutils` and elsewhere.

**Comment by nc-x** [comment]
Seems to compile fine now, most probably due to https://github.com/nim-lang/Nim/commit/f551b72fba30cb70bddfebad9d1398031b5ef620
But IDK if works as expected

**Comment by Araq** [comment]
Thas has been fixed.

**Comment by timotheecour** [comment]
can we reopen this or add explanations on how this was fixed? The original post still gives compile time error:

nim c -r -d:useNimRtl --skipUserCfg --skipParentCfg main.nim
```
Error: VM not allowed to do FFI, see `compiletimeFFI`
```
Likewise when using any proc marked with `{.rtl.}` at compile time, making rtl (and therefore shared libraries) unusable in lots of cases (as soon as a rtl proc is used at compile time, which covers a lot of basic procs; eg simply `import pkg/regex` will fail with `-d:useNimRtl`)
(verified with nim 0.20 or latest devel e7471cebae2a404f3e4239f199f5a0c422484aac)

As explained in https://github.com/nim-lang/Nim/pull/10150#issuecomment-462660376, #10150 doesn't quite fix #8405 because there are still unhandled cases of type mappings in `mapType`; eg `Natural`, `tyRange`, `distinct` aren't mapped, and returning `string` isn't correct currently. I've done more work (in private fork) to improve CTFFI support but as of now nim devel doesn't handle those cases.

The correct fix IMO is to ignore `{.importc.}` pragma when the proc has a body and is used at compile time. I tried defining a macro `rtl` that could be used as a pragma to replace`{.rtl.}` as follows:
```nim
  macro rtl(fun: untyped) =
    result = fun
    let funBody = result[6]
    let ret = result[3][0]
    let isVoid = ret.kind == nnkEmpty # TODO: handle `void` return, auto etc
    let fun0 = $(fun[0])
    let fun2 = copy(fun)
    fun2[0] = ident"funImpl"
    fun2[6] = newEmptyNode()
    fun2[4] = newTree(nnkPragma, newTree(nnkExprColonExpr,ident"importc", newLit fun0))
    let call = newCall(fun2[0])
    for ai in result[3][1..^1]:
      call.add ai[0]
    let retAux = quote do:
      when nimvm:
        `funBody`
      else:
        `fun2`
        `call`

    result[6] = if isVoid:
      newTree(nnkAsgn, ident"result", retAux)
    else:
      retAux
```

which, when used as follows:

```
proc fun2(a0: int; a1 = "bar"): int {.rtl.} =
  1 + 2
```
produces:
```
proc fun2(a0: int; a1 = "bar"): int =
  when nimvm:
    1 + 2
  else:
    proc funImpl(a0: int; a1 = "bar"): int {.importc: "fun2".}
    funImpl(a0, a1)
```

however, that doesn't work (yet, at least, help welcome), for eg `{.rtl.}` is used in places where `std/macros` isn't defined (std/macros itself includes `inclrtl.nim`)

## slimmer rtl
as mentioned in https://github.com/nim-lang/Nim/issues/8405#issuecomment-416009341, I wonder why rtl isn't limited to just the gc parts instead of additional procs scattered in strutils etc; at the very least there should be an option for a `mini_nimrtl` that would do that, as mentioned by @awr1
